snapcraft/tests/repo/test_snaps.py

```diff
@@ -0,0 +1,399 @@
+            if self._is_snap_command(cmd):
+                self.calls.append(cmd)
+                return self._fake_snap_command(cmd, *args, **kwargs)
+            else:
+                return original(cmd, *args, **kwargs)
+
+        check_call_patcher = mock.patch(
+            'snapcraft.internal.repo.snaps.check_call',
+            side_effect=side_effect_check_call)
+        self.mock_call = check_call_patcher.start()
+        self.addCleanup(check_call_patcher.stop)
+
+        check_output_patcher = mock.patch(
+            'snapcraft.internal.repo.snaps.check_output',
+            side_effect=side_effect_check_output)
+        self.mock_call = check_output_patcher.start()
+        self.addCleanup(check_output_patcher.stop)
+
+    def login(self, email):
+        self._email = email
+
+    def _get_snap_cmd(self, snap_cmd):
+        try:
+            snap_cmd_index = snap_cmd.index('snap')
+        except ValueError:
+            return ''
+
+        try:
+            return snap_cmd[snap_cmd_index+1]
+        except IndexError:
+            return ''
+
+    def _is_snap_command(self, cmd):
+        return self._get_snap_cmd(cmd) in ['install', 'refresh', 'whoami']
+
+    def _fake_snap_command(self, cmd, *args, **kwargs):
+        cmd = self._get_snap_cmd(cmd)
+        if cmd == 'install' and not self.install_success:
+            raise subprocess.CalledProcessError(returncode=1, cmd=cmd)
+        elif cmd == 'refresh' and not self.refresh_success:
+            raise subprocess.CalledProcessError(returncode=1, cmd=cmd)
+        elif cmd == 'whoami':
+            return 'email: {}'.format(self._email).encode()
+
+
+class FakeSnapdServer(BaseHTTPRequestHandler):
+
+    def do_GET(self):
+        parsed_url = parse.urlparse(self.path)
+        if parsed_url.path.startswith('/v2/snaps/'):
+            self._handle_snaps(parsed_url)
+        elif parsed_url.path == '/v2/find':
+            self._handle_find(parsed_url)
+        else:
+            self.wfile.write(parsed_url.path.encode())
+
+    def _handle_snaps(self, parsed_url):
+        status_code = 404
+        params = {}
+
+        if parsed_url.path.endswith('/fake-snap'):
+            status_code = 200
+            params = {'channel': 'stable'}
+        elif parsed_url.path.endswith('/fake-snap-stable'):
+            status_code = 200
+            params = {'channel': 'stable'}
+        elif parsed_url.path.endswith('/fake-snap-branch'):
+            status_code = 200
+            params = {'channel': 'candidate/branch'}
+        elif parsed_url.path.endswith('/fake-snap-track-stable'):
+            status_code = 200
+            params = {'channel': 'track/stable'}
+        elif parsed_url.path.endswith('/fake-snap-track-stable-branch'):
+            status_code = 200
+            params = {'channel': 'track/stable/branch'}
+        elif parsed_url.path.endswith('/fake-snap-edge'):
+            status_code = 200
+            params = {'channel': 'edge'}
```

Comment:
This unending stream of elifs caused many problems when we left the fake servers evolve. That's why on the new ones we are using pyramid to route the requests. This is not a lot better, just a little, but it will help for when we are finally able to use https://github.com/pretenders/pretenders

Reply:
I don't like complicating things at the start though.